From f764368a5b708b12d1c3a77b47b4e66e15b11f5a Mon Sep 17 00:00:00 2001
From: Faouaz TENOUTIT <faouaz.tenoutit@intel.com>
Date: Tue, 18 Nov 2014 11:37:46 +0100
Subject: [PATCH] ASoC: Intel: Set GPIO direction before releasing desc

<4>[    4.324468] ------------[ cut here ]------------
<4>[    4.329638] WARNING: CPU: 0 PID: 1 at drivers/gpio/gpiolib.c:208 gpio_ensure_requested+0x42/0x90()
<7>[    4.339652] autorequest GPIO-359
<7>[    4.339655] Modules linked in:
<7>[    4.339663] CPU: 0 PID: 1 Comm: swapper/0 Tainted: G        W    3.14.19-quilt-0053abba #1
<7>[    4.339666] Hardware name: TrekStor SurfTab/ST70408, BIOS ESKU_X64_R_2014_45_2_02_Eng 11/03/2014
<7>[    4.339670]  0000000000000009 ffff8800350d3b00 ffffffff819760f2 ffff8800350d3b48
<7>[    4.339676]  ffff8800350d3b38 ffffffff81077fe3 ffffffff82262e70 0000000000000167
<7>[    4.339682]  ffff88003c05d7e8 0000000000000000 0000000000000001 ffff8800350d3b98
<7>[    4.339688] Call Trace:
<7>[    4.339698]  [<ffffffff819760f2>] dump_stack+0x4e/0x7a
<7>[    4.339706]  [<ffffffff81077fe3>] warn_slowpath_common+0x73/0x90
<7>[    4.339712]  [<ffffffff81078047>] warn_slowpath_fmt+0x47/0x50
<7>[    4.339718]  [<ffffffff813b6122>] gpio_ensure_requested+0x42/0x90
<7>[    4.339724]  [<ffffffff813b6688>] gpiod_direction_output+0x78/0x1b0
<7>[    4.339731]  [<ffffffff8182f52a>] byt_init+0x2ba/0x680

Release the GPIO desc only when all the GPIO's related operations (export, direction, log...) are done.

Signed-off-by: Faouaz TENOUTIT <faouaz.tenoutit@intel.com>
---
 sound/soc/intel/board/byt_cr_dpcm_rt5640.c | 19 +++++++++----------
 1 file changed, 9 insertions(+), 10 deletions(-)

diff --git a/sound/soc/intel/board/byt_cr_dpcm_rt5640.c b/sound/soc/intel/board/byt_cr_dpcm_rt5640.c
index 7379b55..8c4003f 100644
--- a/sound/soc/intel/board/byt_cr_dpcm_rt5640.c
+++ b/sound/soc/intel/board/byt_cr_dpcm_rt5640.c
@@ -563,14 +563,13 @@ static int byt_init(struct snd_soc_pcm_runtime *runtime)
 	desc = devm_gpiod_get_index(codec->dev, NULL, RT5640_GPIO_JD_INT);
 	if (!IS_ERR(desc)) {
 		drvdata->gpios.jd_int_gpio = desc_to_gpio(desc);
-		devm_gpiod_put(codec->dev, desc);
-
 		byt_export_gpio(desc, "JD-int");
 
 		pr_info("%s: GPIOs - JD-int: %d (pol = %d, val = %d)\n",
 			__func__, drvdata->gpios.jd_int_gpio,
 			gpiod_is_active_low(desc), gpiod_get_value(desc));
 
+		devm_gpiod_put(codec->dev, desc);
 	} else {
 		drvdata->gpios.jd_int_gpio = RT5640_GPIO_NA;
 		pr_err("%s: GPIOs - JD-int: Not present!\n", __func__);
@@ -579,14 +578,13 @@ static int byt_init(struct snd_soc_pcm_runtime *runtime)
 	desc = devm_gpiod_get_index(codec->dev, NULL, RT5640_GPIO_JD_INT2);
 	if (!IS_ERR(desc)) {
 		drvdata->gpios.jd_int2_gpio = desc_to_gpio(desc);
-		devm_gpiod_put(codec->dev, desc);
-
 		byt_export_gpio(desc, "JD-int2");
 
 		pr_info("%s: GPIOs - JD-int2: %d (pol = %d, val = %d)\n",
 			__func__, drvdata->gpios.jd_int2_gpio,
 			gpiod_is_active_low(desc), gpiod_get_value(desc));
 
+		devm_gpiod_put(codec->dev, desc);
 	} else {
 		drvdata->gpios.jd_int2_gpio = RT5640_GPIO_NA;
 		pr_warn("%s: GPIOs - JD-int2: Not present!\n", __func__);
@@ -595,8 +593,6 @@ static int byt_init(struct snd_soc_pcm_runtime *runtime)
 	desc = devm_gpiod_get_index(codec->dev, NULL, RT5640_GPIO_JACK_SWITCH);
 	if (!IS_ERR(desc)) {
 		drvdata->gpios.debug_mux_gpio = desc_to_gpio(desc);
-		devm_gpiod_put(codec->dev, desc);
-
 		byt_export_gpio(desc, "debug-mux");
 
 		dir = gpiod_get_direction(desc);
@@ -613,6 +609,8 @@ static int byt_init(struct snd_soc_pcm_runtime *runtime)
 		pr_info("%s: GPIOs - Debug-mux: %d (dir = %d, val = %d)\n",
 			__func__, drvdata->gpios.debug_mux_gpio, dir,
 			gpiod_get_value(desc));
+
+		devm_gpiod_put(codec->dev, desc);
 	} else {
 		drvdata->gpios.debug_mux_gpio = RT5640_GPIO_NA;
 		pr_warn("%s: GPIOs - Debug-mux: Not present!\n", __func__);
@@ -621,14 +619,13 @@ static int byt_init(struct snd_soc_pcm_runtime *runtime)
 	desc = devm_gpiod_get_index(codec->dev, NULL, RT5640_GPIO_JD_BUTTONS);
 	if (!IS_ERR(desc)) {
 		drvdata->gpios.jd_buttons_gpio = desc_to_gpio(desc);
-		devm_gpiod_put(codec->dev, desc);
-
 		byt_export_gpio(desc, "JD-buttons");
 
 		pr_info("%s: GPIOs - JD-buttons: %d (pol = %d, val = %d)\n",
 			__func__, drvdata->gpios.jd_buttons_gpio,
 			gpiod_is_active_low(desc), gpiod_get_value(desc));
 
+		devm_gpiod_put(codec->dev, desc);
 	} else {
 		drvdata->gpios.jd_buttons_gpio = RT5640_GPIO_NA;
 		pr_warn("%s: GPIOs - JD-buttons: Not present!\n", __func__);
@@ -637,18 +634,20 @@ static int byt_init(struct snd_soc_pcm_runtime *runtime)
 	desc = devm_gpiod_get_index(codec->dev, NULL, RT5640_GPIO_I2S_TRISTATE);
 	if (!IS_ERR(desc)) {
 		drvdata->gpios.i2s_tristate_en_gpio = desc_to_gpio(desc);
-		devm_gpiod_put(codec->dev, desc);
 
-		byt_export_gpio(desc, "I2S-Tristate-En");
 		ret = gpiod_direction_output(desc, 0);
 		if (ret)
 			pr_warn("%s: Failed to set direction for GPIO%d (err = %d)!\n",
 				__func__, drvdata->gpios.i2s_tristate_en_gpio,
 				ret);
 
+		byt_export_gpio(desc, "I2S-Tristate-En");
+
 		pr_info("%s: GPIOs - I2S-Tristate-En: %d (pol = %d, val = %d)\n",
 			__func__, drvdata->gpios.i2s_tristate_en_gpio,
 			gpiod_is_active_low(desc), gpiod_get_value(desc));
+
+		devm_gpiod_put(codec->dev, desc);
 	} else {
 		drvdata->gpios.i2s_tristate_en_gpio = RT5640_GPIO_NA;
 		pr_warn("%s: GPIOs - i2s_tristate_en-mux: Not present!\n",
-- 
2.1.1

